cpp/src/join/hash_join.cu

```diff
@@ -87,62 +90,82 @@ std::size_t compute_join_output_size(table_device_view build_table,
   }
 
   auto const probe_nulls = cudf::nullate::DYNAMIC{has_nulls};
-  pair_equality equality{probe_table, build_table, probe_nulls, nulls_equal};
 
-  row_hash hash_probe{probe_nulls, probe_table};
+  auto const row_hash           = cudf::experimental::row::hash::row_hasher{preprocessed_probe};
+  auto const hash_probe         = row_hash.device_hasher(probe_nulls);
   auto const empty_key_sentinel = hash_table.get_empty_key_sentinel();
   make_pair_function pair_func{hash_probe, empty_key_sentinel};
 
-  auto iter = cudf::detail::make_counting_transform_iterator(0, pair_func);
+  auto const row_comparator =
+    cudf::experimental::row::equality::two_table_comparator{preprocessed_probe, preprocessed_build};
+
+  auto const comparator_helper = [&](auto device_comparator) {
+    pair_equality equality{device_comparator};
+
+    auto const iter = cudf::detail::make_counting_transform_iterator(0, pair_func);
```

Comment:
Move `iter` out of this lambda, as it is constructed independently.

Reply:
Also delete `pair_func` and use this:
```suggestion
    auto const iter = cudf::detail::make_counting_transform_iterator(0, make_pair_function{hash_probe, empty_key_sentinel});
```

Reply:
And, in addition (sorry for more request), can you test the approach similar to @davidwendt did in https://github.com/rapidsai/cudf/pull/12900 please? In particular, this: https://github.com/rapidsai/cudf/pull/12900/files#diff-af60c754d22ac70d3ddd68b226bb17ad69657880140ba64d23d5601eb570a690R80-R87. 

I'm curious if we can save compile time by switching away from `counting_transform_iterator`.

Reply:
As @davidwendt has commented in the PR as well as offline, it's not `counting_transform_iterator` that's adding to the compile-time, the issue is with using it combined with some `thrust` calls. In this instance, it's being passed to `cuco`.

@PointKernel @davidwendt will have better ideas if `cuco` has the same inlining problems as `thrust`, or if it's okay to use fancy iterators.

Reply:
At first glance, `cuco` does not appear to have an aggressive amount of inlining of the comparator like thrust/cub does.
Even so, I don't see how the results of the comparator can be prefilled into a temporary buffer for `cuco` to use in this case like we could in the 12900 examples.
Regardless, I do want to encourage this line of thinking/investigating/reviewing. 😃

Reply:
The input can probably be prefilled into a temporary buffer though, by materializing the count and transform operations. It will be excess memory usage because like you said, it's not the results that are being prefilled.